### ISSUE TITLE
Distorted on skins besides Estuary

### ISSUE THREAD

**Issue by JayDee696969** [doko-desuka/plugin.program.orderfavourites#10]
Hey, just a quick question. Is this addon only compatible with the default Estuary skin? Other skins it is all out of alignment? Its fine on Estuary, but for example it looks like this with Xonfluence skin.
<img width="794" alt="Screenshot_36" src="https://user-images.githubusercontent.com/45118866/211429566-e06dbdf6-9541-49b4-a881-2074259ee3c1.png">

**Comment by doko-desuka** [comment]
Wow, that looks bad.

Yes, I only tested it with Estuary, the default Kodi skin. All of the UI layout is in this XML file here:
https://github.com/doko-desuka/plugin.program.orderfavourites/blob/master/resources/skins/Default/1080i/CustomFavouritesDialog.xml

I wouldn't know how to fix that.

**Comment by JayDee696969** [comment]
Ok, thanks.

**Comment by M-Borsch** [comment]
Forked a version that works with Confluence: https://github.com/m-borsch

**Comment by doko-desuka** [comment]
@M-Borsch hi, that looks interesting. Did you find if it can only work with one skin at a time, or is there a generic way to make it work with both?

Also, after using the original for a while, the ordering method started to annoy me.  
I tested changing it to swapping (that is, instead of item A being inserted in front of B, you just have A swap places with B), and I found it much less confusing and it was faster to organize too. I'll see if I can push it to master, at least with a setting in case someone wants to revert to the old way of ordering.

**Comment by M-Borsch** [comment]
@doko-desuka - Hi. The changes were just layout changes all done in CustomFavouritesDialog.xml. Since the values are hard coded, I simply reviewed the layout design and made some edits. I am using the Confluence skin so made them such that they work properly for it. I did not test the changes in other skins yet but it should work for any skin. As you can see from my other comment, I also packaged it up into a Repository. If I have some time, I may see how it looks on other skins.

As for your sort/swapping change. I believe that you don't want to "swap" A and B (as that would not make sense when trying to insert from the bottom of the list?) - but rather I think that you mean place it at B (meaning "after" B i.e. "leaving B in front of A"?). If that is what you mean - that is just a minor change.

The pure "swapping" of A and B would not make sense in the case where you add a new favourite to an alphabetically sorted list since it defaults to the bottom of the list. You then want to insert it at location B either "before" (current version) or "after" B (what I believe you are considering). 

Keep me posted on what you decide to do.

**Comment by doko-desuka** [comment]
> As for your sort/swapping change. I believe that you don't want to "swap" A and B (as that would not make sense when trying to insert from the bottom of the list?) - but rather I think that you mean place it at B (meaning "after" B i.e. "leaving B in front of A"?). If that is what you mean - that is just a minor change.

No, I mean that A and B, wherever they are in the favourites items, will take the place of one another.  
So take the following favourites grid:
```
00 AA 11 22 33
44 55 BB 66 77
```

When re-ordering, the old method would move AA in front of B:
```
00 11 22 33 44
55 BB AA 66 77
```

Now if instead they are swapped, you get this:
```
00 BB 11 22 33
44 55 AA 66 77
```

This is so much more intuitive, because on the Kodi Favourites screen there's a big list with rows of 4 items. If you have a lot of favourites, you end up getting used to their **visual** layout, that is, how some items are on the extreme-left column, the two columns in the middle, and the extreme-right column.  
If you use the old method of "remove AA and re-insert it around BB" (no matter if in front of or behind it), potentially so many items will have their visual order affected. With swapping, only AA and BB change.  
It's easier to understand when trying it out, I'll pull it to master.

**Comment by doko-desuka** [comment]
Ok, 1.3.0 is ready with this new swap behavior.  
You can find it in the Releases page: https://github.com/doko-desuka/plugin.program.orderfavourites/releases

Edit: I also added a link to your fork in the [Kodi forums thread](https://forum.kodi.tv/showthread.php?tid=351844&page=2), thanks for working on it.

**Comment by M-Borsch** [comment]
@doko-desuka - ok now I understand what you mean. When I get a minute, and since we don't have a "switch" option, I will likely just create another "forked" version - but will call this new Program addon something like "Order Favourites - Swap" for you to try out. That way folks can try/use either or both.

**Comment by M-Borsch** [comment]
> @M-Borsch hi, that looks interesting. Did you find if it can only work with one skin at a time, or is there a generic way to make it work with both?
> 
@doko-desuka - I had an opportunity to test out this installed and changing to a number of skins and the answer is "yes" it works on all the skins that I tested (Aeon Nox: SiLVO,  Confluence, Estouchy, and Estuary).

**Comment by M-Borsch** [comment]
> @doko-desuka - ok now I understand what you mean. When I get a minute, and since we don't have a "switch" option, I will likely just create another "forked" version - but will call this new Program addon something like "Order Favourites - Swap" for you to try out. That way folks can try/use either or both.

@doko-desuka - Ok I have added the new Program Addon and created a new version of the repository (V2.0) that serves up both versions of the Addons.

Here they are loaded using the Repository on Confluence:

<img width="480" alt="Order Favourites" src="https://github.com/doko-desuka/plugin.program.orderfavourites/assets/154018836/d9cba732-a763-48c5-886a-73422525cca2">

**Comment by M-Borsch** [comment]
I have revamped and renamed the Repo that delivers new versions of the 2 options:
Order Favourites - Insert
Order Favourites - Swap.

Here is the updated repository: https://[m-borsch.github.io/MB-KODI/](https://m-borsch.github.io/MB-KODI/)

The icons and descriptions have been changed to make it obvious what each version does:
<img width="800" alt="Kodi-2addons" src="https://github.com/doko-desuka/plugin.program.orderfavourites/assets/154018836/f320ee36-fcb5-4c91-a13d-baf4de9669ba">

Also, I have added the icons and hints on each of the editing screens (under the "Restore.." button):
<img width="234" alt="image" src="https://github.com/doko-desuka/plugin.program.orderfavourites/assets/154018836/62763531-40f6-4f5d-b612-ad5064a5bdde">

Code is located at: https://github.com/m-borsch

**Comment by M-Borsch** [comment]
Here is what the installs look like from the Repository (Insert version):
<img width="466" alt="Kodi-plugin-authors" src="https://github.com/doko-desuka/plugin.program.orderfavourites/assets/154018836/b7353569-aef6-444e-8826-021ca5a2d392">

Here is the Addon information screen (Swap version):
<img width="662" alt="Kodi-plugin-description" src="https://github.com/doko-desuka/plugin.program.orderfavourites/assets/154018836/30e9d4d1-8e09-48aa-a1ea-c4fefce78b8a">

**Comment by doko-desuka** [comment]
Thanks. Putting the info on the icon communicates it well, but wouldn't it be cleaner to add it as a setting? This way the user can set it to their preferred method, and it can be made into a single plugin instead of two.

If you need some reference, you can save this code below as a file named `settings.xml` inside of `plugin.program.orderfavourites/resources/`:
```xml
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<settings>
    <category label="Behavior">
        <setting id="reorderingMethod" label="Reordering Method" type="enum" default="0"
                 values="Swap A and B|Place A before B|Place A ahead of B"/>
    </category>
</settings>

```

After that, the function `doSelect` in script `default.py` can be replaced by the one below, which reacts differently depending on that setting. Be careful with the indentation of the first line after you paste it:
```python
    def doSelect(self):
        selectedPosition = self.panel.getSelectedPosition()
        if self.indexFrom == None:
            # Selecting a new item to reorder.
            self.indexFrom = selectedPosition
            self.panel.getSelectedItem().setProperty('selected', '1')
        else:
            # Something was already selected, so do the reodering.
            if self.indexFrom != selectedPosition:
                self.allItems[self.indexFrom].setProperty('selected', '')

                # Reorder the two distinct items in a specific way:
                reorderingMethod = ADDON.getSetting('reorderingMethod')
                # If using the swap mode, or if the items are direct neighbors, then
                # just swap them.
                if reorderingMethod == '0' \
                   or (self.indexFrom == (selectedPosition + 1)) \
                   or (self.indexFrom == (selectedPosition - 1)):
                    # Swap A and B.
                    self.allItems[self.indexFrom], self.allItems[selectedPosition] = (
                        self.allItems[selectedPosition], self.allItems[self.indexFrom]
                    )
                else:
                    itemFrom = self.allItems.pop(self.indexFrom)
                    if reorderingMethod == '1':
                        # Place A behind B.
                        # In case A is at some point BEHIND of B, reduce
                        # one index because popping A caused the list to shrink.
                        if self.indexFrom < selectedPosition:
                            selectedPosition = selectedPosition - 1
                    else:
                        # Place A ahead of B (the original ordering method).
                        # In case A is at some point AHEAD of B, move up
                        # one index because .insert() always puts it behind.
                        if self.indexFrom > selectedPosition:
                            selectedPosition = selectedPosition + 1
                    self.allItems.insert(selectedPosition, itemFrom)

                # Reset the selection state.
                self.isDirty = True
                self.indexFrom = None

                # Commit the changes to the UI, and highlight item A.
                self.panel.reset()
                self.panel.addItems(self.allItems)
                self.panel.selectItem(selectedPosition)
            else: # User reselected the item, so just unmark it.
                self.indexFrom = None
                self.panel.getSelectedItem().setProperty('selected', '')

```
...and a disclaimer that I didn't test it, so I wouldn't be surprised if there was a syntax or logic error in there.

**Comment by M-Borsch** [comment]
> Thanks. Putting the info on the icon communicates it well, but wouldn't it be cleaner to add it as a setting? This way the user can set it to their preferred method, and it can be made into a single plugin instead of two.
> 
> If you need some reference, you can save this code below as a file named `settings.xml` inside of `plugin.program.orderfavourites/resources/`:
> 
> ```
> <?xml version="1.0" encoding="utf-8" standalone="yes"?>
> <settings>
>     <category label="Behavior">
>         <setting id="reorderingMethod" label="Reordering Method" type="enum" default="0"
>                  values="Swap A and B|Place A before B|Place A ahead of B"/>
>     </category>
> </settings>
> ```
> 
> After that, the function `doSelect` in script `default.py` can be replaced by the one below, which reacts differently depending on that setting. Be careful with the indentation of the first line after you paste it:
> 
> ```python
>     def doSelect(self):
>         selectedPosition = self.panel.getSelectedPosition()
>         if self.indexFrom == None:
>             # Selecting a new item to reorder.
>             self.indexFrom = selectedPosition
>             self.panel.getSelectedItem().setProperty('selected', '1')
>         else:
>             # Something was already selected, so do the reodering.
>             if self.indexFrom != selectedPosition:
>                 self.allItems[self.indexFrom].setProperty('selected', '')
> 
>                 # Reorder the two distinct items in a specific way:
>                 reorderingMethod = ADDON.getSetting('reorderingMethod')
>                 # If using the swap mode, or if the items are direct neighbors, then
>                 # just swap them.
>                 if reorderingMethod == '0' \
>                    or (self.indexFrom == (selectedPosition + 1)) \
>                    or (self.indexFrom == (selectedPosition - 1)):
>                     # Swap A and B.
>                     self.allItems[self.indexFrom], self.allItems[selectedPosition] = (
>                         self.allItems[selectedPosition], self.allItems[self.indexFrom]
>                     )
>                 else:
>                     itemFrom = self.allItems.pop(self.indexFrom)
>                     if reorderingMethod == '1':
>                         # Place A behind B.
>                         # In case A is at some point BEHIND of B, reduce
>                         # one index because popping A caused the list to shrink.
>                         if self.indexFrom < selectedPosition:
>                             selectedPosition = selectedPosition - 1
>                     else:
>                         # Place A ahead of B (the original ordering method).
>                         # In case A is at some point AHEAD of B, move up
>                         # one index because .insert() always puts it behind.
>                         if self.indexFrom > selectedPosition:
>                             selectedPosition = selectedPosition + 1
>                     self.allItems.insert(selectedPosition, itemFrom)
> 
>                 # Reset the selection state.
>                 self.isDirty = True
>                 self.indexFrom = None
> 
>                 # Commit the changes to the UI, and highlight item A.
>                 self.panel.reset()
>                 self.panel.addItems(self.allItems)
>                 self.panel.selectItem(selectedPosition)
>             else: # User reselected the item, so just unmark it.
>                 self.indexFrom = None
>                 self.panel.getSelectedItem().setProperty('selected', '')
> ```
> 
> ...and a disclaimer that I didn't test it, so I wouldn't be surprised if there was a syntax or logic error in there.

Agreed. However, originally I had attempted to isolate any changes to quick layout changes to address issues with the other skins not rendering properly.

I will try to carve out some time and see if I can create a test version and introduce the "option" approach integrated into a single addon.

**Comment by M-Borsch** [comment]
@doko-desuka - thanks for the suggestion. I have imported the code and cleaned it up a bit. I have released a new Program Addon "Order Favourites" as well as updated the repository to take out the 2 previous versions and replaced with this new version that uses the configuration panel to set the action to be taken (one of the 3 options).

Here is the new Addon Description:
<img width="667" alt="OrderFavourites-Description" src="https://github.com/doko-desuka/plugin.program.orderfavourites/assets/154018836/56afbaad-009e-4deb-8296-72c1d2935052">

Here is the new Configuration panel:
<img width="623" alt="OrderFavourites-Configure" src="https://github.com/doko-desuka/plugin.program.orderfavourites/assets/154018836/7d4cb918-b9a6-4f5d-9e16-36a81ad60f63">

Here is the updated repository MB-KODI (3.1): https://[m-borsch.github.io/MB-KODI/](https://m-borsch.github.io/MB-KODI/)

Code is located at: https://github.com/m-borsch

**Comment by doko-desuka** [comment]
I downloaded Xonfluence, opened the dialog, found the problems (lack of default values, lack of explicit widths), and fixed it on 1.4.0 which resolves this issue.  

You can find 1.4.0 and up on the Releases page:  
https://github.com/doko-desuka/plugin.program.orderfavourites/releases

@M-Borsch could you please rename your repositories to avoid having the exact same name as OrderFavourites? There are two: one named the same as this one, and another named OrderFavourites.  
I didn't think about it before, but when someone looks for this it'll lead to confusion, like "which is the original, which is the fork"?   When people fork someone else's plugin they usually completely change the identity. See [Elementum](https://github.com/elgatito/plugin.video.elementum#what-it-is), which is a fork of Quasar, which is a fork of Pulsar, for example. Cheers

**Comment by M-Borsch** [comment]
> I downloaded Xonfluence, opened the dialog, found the problems (lack of default values, lack of explicit widths), and fixed it on 1.4.0 which resolves this issue.
> 
> You can find 1.4.0 and up on the Releases page: https://github.com/doko-desuka/plugin.program.orderfavourites/releases
> 
> @M-Borsch could you please rename your repositories to avoid having the exact same name as OrderFavourites? There are two: one named the same as this one, and another named OrderFavourites. I didn't think about it before, but when someone looks for this it'll lead to confusion, like "which is the original, which is the fork"? When people fork someone else's plugin they usually completely change the identity. See [Elementum](https://github.com/elgatito/plugin.video.elementum#what-it-is), which is a fork of Quasar, which is a fork of Pulsar, for example. Cheers

Yes, of course - I will rename them all to avoid any confusion.
